File: modules/assets/src/main/java/com/logistimo/assets/entity/AssetStatus.java
/*
 * Copyright © 2017 Logistimo.
 *
 * This file is part of Logistimo.
 *
 * Logistimo software is a mobile & web platform for supply chain management and remote temperature monitoring in
 * low-resource settings, made available under the terms of the GNU Affero General Public License (AGPL).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General
 * Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 * You can be released from the requirements of the license by purchasing a commercial license. To know more about
 * the commercial license, please contact us at opensource@logistimo.com
 */

package com.logistimo.assets.entity;

import com.logistimo.dao.JDOUtils;


import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.jdo.annotations.IdentityType;
import javax.jdo.annotations.NotPersistent;
import javax.jdo.annotations.PersistenceCapable;
import javax.jdo.annotations.Persistent;
import javax.jdo.annotations.PrimaryKey;

/**
 * Created by charan on 27/11/15.
 */
@PersistenceCapable(identityType = IdentityType.APPLICATION, detachable = "true")
public class AssetStatus implements IAssetStatus {

  /**
   * Hash based on Asset Id, Status Type, Monitoing point Id, Sensor Id.
   */
  @PrimaryKey
  @Persistent
  private Long id;

  /**
   * Asset Id
   */
  @Persistent
  private Long assetId;

  /**
   * Monitoring point Id
   */
  @Persistent
  private Integer mpId;

  /**
   * Sensor Id . Sent for device alarms, Sensor device.
   */
  @Persistent
  private String sId;

  /**
   * Alarm type. The following alarm types.
   * - Device status
   * - Temperature status
   * - Battery status
   * - Activity status
   * - External Sensor status
   * - Connection status , Transmitter not connected.
   */
  @Persistent
  private Integer type;

  @Persistent
  private Integer status;

  @Persistent
  private Integer abnStatus;

  @Persistent
  private Float tmp;

  @Persistent
  private Date ts;

  @Persistent(defaultFetchGroup = "true", mappedBy = "assetStatusId", dependentElement = "true")
  private List<AssetAttribute> attributes;

  @NotPersistent
  private List<String> tags;

  @Override
  public Long getId() {
    return id;
  }

  @Override
  public void setId(Long id) {
    this.id = id;
  }

  @Override
  public Long getAssetId() {
    return assetId;
  }

  @Override
  public void setAssetId(Long assetId) {
    this.assetId = assetId;
  }

  @Override
  public Integer getMpId() {
    return mpId;
  }

  @Override
  public void setMpId(Integer mpId) {
    this.mpId = mpId;
  }

  @Override
  public Integer getStatus() {
    return status;
  }

  @Override
  public void setStatus(Integer status) {
    this.status = status;
  }

  @Override
  public Integer getAbnStatus() {
    return abnStatus;
  }

  @Override
  public void setAbnStatus(Integer abnStatus) {
    this.abnStatus = abnStatus;
  }

  @Override
  public List<String> getTags() {
    return this.tags;
  }

  @Override
  public void setTags(List<String> tags) {
    this.tags = tags;
  }

  @Override
  public Float getTmp() {
    return tmp;
  }

  @Override
  public void setTmp(Float tmp) {
    this.tmp = tmp;
  }

  @Override
  public Date getTs() {
    return ts;
  }

  @Override
  public void setTs(Date ts) {
    this.ts = ts;
  }

  @Override
  public void init(Long assetId, int mpId, String sId, int type, int st, int aSt, double tmp,
                   int time, List<IAssetAttribute> attrs, List<String> tags) {
    this.assetId = assetId;
    this.mpId = mpId;
    this.sId = sId;
    this.type = type;
    this.status = st;
    this.abnStatus = aSt;
    this.tmp = new Float(tmp);
    this.ts = new Date();
    this.ts.setTime(time * 1000l);
    if (this.attributes == null) {
      this.attributes = new ArrayList<>(1);
    }
    if (attrs != null) {
      for (IAssetAttribute iAssetAttribute : attrs) {
        this.attributes.add((AssetAttribute) iAssetAttribute);
      }
    }
    this.tags = tags;
    this.id = JDOUtils.createAssetStatusKey(assetId, mpId, sId, type);
  }

  @Override
  public List<AssetAttribute> getAttributes() {
    return attributes;
  }

  @Override
  public void setAttributes(List<? extends IAssetAttribute> attributes) {
    if (this.attributes == null) {
      this.attributes = new ArrayList<>(1);
    } else {
      this.attributes.clear();
    }
    for (IAssetAttribute iAssetAttribute : attributes) {
      this.attributes.add((AssetAttribute) iAssetAttribute);
    }

  }

  @Override
  public void addAttribute(IAssetAttribute assetAttribute) {
    if (attributes == null) {
      attributes = new ArrayList<>(1);
    }
    attributes.add((AssetAttribute) assetAttribute);
  }

  @Override
  public Integer getType() {
    return type;
  }

  @Override
  public void setType(Integer type) {
    this.type = type;
  }

  @Override
  public String getsId() {
    return sId;
  }

  @Override
  public void setsId(String sId) {
    this.sId = sId;
  }

  @Override
  public String toString() {
    return "AssetStatus{" +
        "assetId=" + assetId +
        ", mpId=" + mpId +
        ", sId='" + sId + '\'' +
        ", type=" + type +
        ", status=" + status +
        ", abnStatus=" + abnStatus +
        ", tmp=" + tmp +
        ", ts=" + ts +
        '}';
  }
}

File: modules/web/src/main/webapp/v2/plugins/reports/asset/assetstatus/assetstatus.js
/*
 * Copyright © 2017 Logistimo.
 *
 * This file is part of Logistimo.
 *
 * Logistimo software is a mobile & web platform for supply chain management and remote temperature monitoring in
 * low-resource settings, made available under the terms of the GNU Affero General Public License (AGPL).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General
 * Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 * You can be released from the requirements of the license by purchasing a commercial license. To know more about
 * the commercial license, please contact us at opensource@logistimo.com
 */

/**
 * Created by Mohan Raja on 23/02/17.
 */

registerWidget('aas', 'rpt-asset-status', 'Assets', 'Asset status','asset/assetstatus');

(function () {
    'use strict';

    var reportType = 'aas';

    reportsPluginCore.directive('rptAssetStatus', function () {
        return {
            restrict: 'E',
            templateUrl: 'plugins/reports/asset/assetstatus/assetstatus.html'
        };
    });

    reportsPluginCore.controller('rptAssetStatusController', ReportAssetStatusController);

    ReportAssetStatusController.$inject = ['$scope', '$timeout','reportsServiceCore','domainCfgService'];

    function ReportAssetStatusController($scope, $timeout, reportsServiceCore,domainCfgService) {

        AssetReportController.call(this, $scope, $timeout, getData);

        $scope.cType = "mscolumn2d";
        $scope.cOptions.exportFileName = "AssetStatus" + "_" + FormatDate_DD_MM_YYYY($scope.today);

        domainCfgService.getAssetWorkingStatusSysCfg().then(function(data){
            var status = data.data;
            for(var s in status) {
                $scope.primaryMetric.push({name: status[s], value: s});
            }
        });

        $scope.secondaryMetric.push({name: 'Number of times assets changed to this state', value: '0'});

        $scope.metricHeadings.push({name: $scope.resourceBundle['overview'], code: "ot"});
        $scope.metricHeadings.push({name: $scope.resourceBundle['by.upper']+" "+$scope.resourceBundle['locations.lower'], code: "rt"});
        $scope.metricHeadings.push({name: $scope.resourceBundle['by.upper']+" "+$scope.resourceBundle['manufacturer.lower'], code: "mnt"});
        $scope.metricHeadings.push({name: $scope.resourceBundle['by.upper']+" "+$scope.resourceBundle['model.lower'], code: "amt"});

        $scope.downloadAsCSV = function (daily) {
            if(daily) {
                var data = $scope.dtableData;
                var heading = $scope.dtableHeading;
            } else {
                data = $scope.tableData;
                heading = $scope.tableHeading;
            }
            var fileName = "Asset_Status" + formatDate2Url($scope.filter.from) +"_"+ formatDate2Url($scope.filter.to);
            $scope.exportAsCSV(data, heading, fileName);
        };

        $scope.downloadTableAsCSV = function () {
            var data = $scope.tableData;
            var heading = $scope.tableCSVHeading;
            var fileName = "Asset_Status" + formatDate2Url($scope.filter.from) +"_"+ formatDate2Url($scope.filter.to);
            $scope.exportAsCSV(data, heading, fileName, $scope.tableSeriesNo);
        };

        $scope.$watch("metrics.primary",function(newValue,oldValue){
            if(newValue != oldValue) {
                if($scope.activeMetric == 'ot'){
                    setChartData();
                } else{
                    setTableMeta();
                }
            }
        });

        $scope.$watch("metrics.secondary",function(newValue,oldValue){
            if(newValue != oldValue) {
                if($scope.activeMetric == 'ot'){
                    setChartData();
                } else {
                    setTableMeta();
                }
            }
        });

        $scope.applyFilter();

        function setTableMeta() {
            $scope.tableSeriesNo = parseInt($scope.metrics.secondary) * $scope.primaryMetric.length
                + parseInt($scope.metrics.primary * 1);
            $scope.tableMetric = $scope.primaryMetric[$scope.metrics.primary - 1].name
                + " - " + $scope.secondaryMetric[$scope.metrics.secondary].name;
        }

        function getData() {
            var selectedFilters = $scope.populateFilters();
            selectedFilters['type'] = reportType;

            if(selectedFilters['level'] == "d") {
                $scope.dLoading = true;
            } else {
                $scope.loading = true;
                $scope.cData = $scope.cLabel = $scope.chartData = undefined;
            }

            if($scope.activeMetric == 'ot') {
                $scope.showLoading();
                reportsServiceCore.getReportData(angular.toJson(selectedFilters)).then(function (data) {
                    $scope.noData = true;
                    if (checkNotNullEmpty(data.data)) {
                        var chartData = angular.fromJson(data.data);
                        chartData = sortByKeyDesc(chartData, 'label');
                        if (selectedFilters['level'] == "d") {
                            $scope.allDChartData = chartData;
                        } else {
                            $scope.allChartData = chartData;
                        }
                        setChartData(true, chartData, selectedFilters.level);
                        $scope.noData = false;
                    }
                }).catch(function error(msg) {
                }).finally(function () {
                    if (selectedFilters['level'] == "d") {
                        $scope.dLoading = false;
                    } else {
                        $scope.loading = false;
                    }
                    $scope.hideLoading();
                });
            } else {
                selectedFilters['viewtype'] = $scope.activeMetric;
                selectedFilters['s'] = $scope.size;
                selectedFilters['o'] = $scope.offset;
                $scope.noData = true;
                $scope.tableHideNext = false;
                $scope.showLoading();
                reportsServiceCore.getReportBreakdownData(angular.toJson(selectedFilters)).then(function (data) {
                    if (checkNotNullEmpty(data.data)) {
                        var breakdownData = angular.fromJson(data.data);

                        setTableMeta();
                        $scope.tableCaption = $scope.getReportCaption(true);

                        $scope.tableHeading = [];
                        $scope.tableHeading = angular.copy(breakdownData.headings);
                        $scope.updateTableHeading($scope.tableHeading,$scope.getReportDateFormat());
                        $scope.tableCSVHeading = breakdownData.headings;
                        $scope.updateTableHeading($scope.tableCSVHeading, "yyyy-MM-dd");
                        $scope.tableData = sortObject(breakdownData.table);
                        formatReportTableData($scope.tableData);
                        $scope.tableDataLength = Object.keys($scope.tableData).length;
                        $scope.noData = false;
                    } else if ($scope.tableDataLength == $scope.size) {
                        $scope.tableHideNext = true;
                        $scope.noData = false;
                        $scope.offset -= $scope.size;
                    }
                }).catch(function error(msg) {
                }).finally(function(){
                    $scope.loading = false;
                    $scope.hideLoading();
                });
            }
        }

        function setChartData(localData, chartData, level) {
            if(!localData) {
                $scope.loading = true;
                chartData = angular.copy($scope.allChartData);
            }
            var linkDisabled = level == "d" || $scope.filter.periodicity == "d";
            var cLabel = getReportFCCategories(chartData,$scope.getReportDateFormat(level));
            var compareFields = [];
            angular.forEach(chartData, function (d) {
                if (compareFields.indexOf(d.value[0].value) == -1) {
                    compareFields.push(d.value[0].value);
                }
            });
            var cData = [];
            for (var i = 0; i < compareFields.length; i++) {
                cData[i] = getReportFCSeries(chartData,
                    $scope.metrics.secondary * $scope.primaryMetric.length + $scope.metrics.primary * 1, compareFields[i],
                    "column2d", linkDisabled, 0);
            }
            $scope.cOptions.caption = $scope.primaryMetric[$scope.metrics.primary - 1].name
                + " - " + $scope.secondaryMetric[$scope.metrics.secondary].name;
            $scope.cOptions.subcaption = $scope.getReportCaption();
            $scope.cOptions.yAxisName = $scope.cOptions.caption;
            if($scope.filter.periodicity != "m" && cLabel.length > 10) {
                $scope.cOptions.rotateLabels = "1";
            } else {
                $scope.cOptions.rotateLabels = undefined;
            }
            if (level == "d") {
                $scope.dcData = cData;
                $scope.dcLabel = cLabel;
                $scope.dchartData = chartData;

                $scope.dtableCaption=undefined;
                $scope.dtableHeading = [];
                $scope.dtableHeading.push("Date");
                if(checkNotNullEmpty(compareFields[0])) {
                    angular.forEach(compareFields, function (d) {
                        $scope.dtableHeading.push(d);
                    });
                } else {
                    $scope.dtableHeading.push($scope.cOptions.caption);
                }
                $scope.dtableData = getReportTableData(cData, cLabel);
                $scope.dtableDataLength = Object.keys($scope.dtableData).length;
                $scope.dcOptions = angular.copy($scope.cOptions);
                if(cLabel.length > 10) {
                    $scope.dcOptions.rotateLabels = "1";
                }
                $scope.dcOptions.rotateValues = "1";
                $scope.dcOptions.caption = $scope.getBreakDownCaption();
                $scope.dcOptions.subcaption = undefined;
                $scope.dtableMetric = $scope.dcOptions.caption;
            } else {
                $scope.cData = cData;
                $scope.cLabel = cLabel;
                $scope.chartData = chartData;
                $scope.dcData = $scope.dcLabel = $scope.dchartData = undefined;

                $scope.tableCaption=$scope.cOptions.subcaption;
                $scope.tableHeading = [];
                $scope.tableHeading.push("Date");
                if(checkNotNullEmpty(compareFields[0])) {
                    $scope.tableMetric = $scope.cOptions.caption;
                    angular.forEach(compareFields, function (d) {
                        $scope.tableHeading.push(d);
                    });
                } else {
                    $scope.tableMetric = undefined;
                    $scope.tableHeading.push($scope.cOptions.caption);
                }
                $scope.tableData = getReportTableData(cData, cLabel,$scope.getReportDateFormat());
                $scope.tableDataLength = Object.keys($scope.tableData).length;
            }
            if(!localData) {
                $timeout(function () {
                    $scope.loading = false;
                }, 200);
            }
        }
    }
})();

